plasmapy/plasma/grids.py

```diff
@@ -289,81 +250,96 @@ def dax2(self):
-
-    def _validate(self):
-        """
-        Checks to make sure that the grid parameters are
-        consistent with the coordinate system and units selected
-
-        """
-        return True
-
-    def _load_grid(self, grid: Union[np.ndarray, u.Quantity], units=None):
+    def _load_grid(
+        self, pts0: u.Quantity, pts1: u.Quantity, pts2: u.Quantity, **kwargs,
+    ):
         """
         Initialize the grid object from a user-supplied grid
 
         Parameters
         ----------
-        grid : np.ndarray or u.Quantity array, shape (n0, n1, n2, 3)
-            A grid of positions in 3D space. If an np.ndarray is provided,
-            the units keyword must be set to provide units for each dimension.
-            If a u.Quantity array is provided, its units will be used.
+        grid{0,1,2} : u.Quantity array, shape (n0, n1, n2)
+            Grids of coordinate positions.
 
-        units : A u.core.Unit object or a list of three of the same
-            Units to be applied to each dimension. If only one unit is given,
-            the same unit will be applied to all dimensions.
+        **kwargs: u.Quantity array, shape (n0, n1, n2)
+            Quantities defined on the grid
 
         Returns
         -------
         None.
 
         """
 
-        # Determine units from grid or units keyword (in that order)
-        if hasattr(grid, "unit"):
-            self._units = [grid.unit] * 3
-        elif units is not None:
-            self._units = units
-        else:
+        # Validate input
+        if not (pts0.shape == pts1.shape and pts0.shape == pts2.shape):
             raise ValueError(
-                "Input grid must be either a u.Quantity array "
-                "or an np.ndarray with units provided in the "
-                " units keyword. Type of grid provided was "
-                f"{type(grid)}."
+                "Provided arrays of grid points are of unequal "
+                f"shape: pts0 = {pts0.shape}, "
+                f"pts1 = {pts1.shape}, "
+                f"pts2 = {pts2.shape}."
             )
 
-        # If quantity array was given, strip units
-        if hasattr(grid, "value"):
-            self._grid = grid.value
+        self.is_uniform_grid = _detect_is_uniform_grid(pts0, pts1, pts2)
+
+        # Create dataset
+        self.ds = xr.Dataset()
+
+        self.ds.attrs["axis_units"] = [pts0.unit, pts1.unit, pts2.unit]
+        if self.is_uniform_grid:
+            self.ds.coords["ax0"] = pts0[:, 0, 0]
+            self.ds.coords["ax1"] = pts1[0, :, 0]
+            self.ds.coords["ax2"] = pts2[0, 0, :]
+
         else:
-            self._grid = grid
+            mdx = pd.MultiIndex.from_arrays(
+                [pts0.flatten(), pts1.flatten(), pts2.flatten()],
+                names=["ax0", "ax1", "ax2"],
+            )
+            self.ds.coords["ax"] = mdx
+
+        # Add quantities
+        for qk in kwargs.keys():
+            q = kwargs[qk]
+
+            self.add_quantity(qk, q)
```

Comment:
I'm having trouble concentrating on this right now, but I'm guessing this is the meat and potatoes of the new idea? I don't see anything immediately wrong with it, but I'll have to take a closer look. I take it the `is_uniform_grid` expects the `numpy.meshgrid` kind of arrays?

Reply:
Yes the idea is that the data can be structured in two ways: 

Uniform grids: Quantities on the grid have shapes [nx,ny,nz], normal 1D coordinate axes are stored. The relationship between these axes and the grid points is as in np.meshgrid, as you said.

Non-uniform grids: Treated as completely unstructed data: grids are flattened to 1D and a MultiIndex coordinate is provided whose three elements give the x,y,z positions of each data point.

Reply:
I'll have to run this through a debugger and get a feel for how the MultiIndex variant works. The idea sounds reasonable though!